crates/polars-plan/src/client/mod.rs

```diff
@@ -0,0 +1,68 @@
+mod dsl;
+
+use polars_core::error::{polars_err, PolarsResult};
+use polars_io::path_utils::is_cloud_url;
+
+use crate::dsl::Expr;
+use crate::plans::options::SinkType;
+use crate::plans::{DslFunction, DslPlan, FunctionNode};
+
+/// Assert that the given [`DslPlan`] is eligible to be executed on Polars Cloud.
+pub fn assert_cloud_eligible(dsl: &DslPlan) -> PolarsResult<()> {
+    let mut expr_stack = vec![];
+    for plan_node in dsl.into_iter() {
+        match plan_node {
+            DslPlan::MapFunction {
+                function: DslFunction::FunctionNode(function),
+                ..
+            } => match function {
+                FunctionNode::Opaque { .. } => return ineligible_error("contains opaque function"),
+                #[cfg(feature = "python")]
+                FunctionNode::OpaquePython { .. } => {
+                    return ineligible_error("contains Python function")
+                },
+                _ => {},
+            },
+            #[cfg(feature = "python")]
+            DslPlan::PythonScan { .. } => return ineligible_error("contains Python scan"),
+            DslPlan::GroupBy { apply: Some(_), .. } => {
+                return ineligible_error("contains Python function in group by operation")
```

Comment:
We should also check the expressions passed as aggregations.

Reply:
That is already checked - this branch is just to error on `map_groups`. If it's not a `map_groups`, it will skip this branch and go to the branch that checks the nested expressions.

I added a test for this case!